### PR TITLE
[dask-on-ray] Add better Dask-on-Ray example, and detail custom shuffle optimization.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR is for a general Dask-on-Ray update. Merging this PR is blocked by the custom shuffle optimization PR being merged.

### CODE REVIEW

File: doc/source/dask-on-ray.rst
Comment (ericl) [inline doc/source/dask-on-ray.rst:91]: Can we enable these by default on ray.init()?

Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:91]: We can, although I think that would be less good for a few reasons:

1. Having the core `ray.init()` path reach out and set Dask configs seems like a poor separation of concerns, and may be odd for those using both Dask and Ray but _not_ trying to use Dask-_on_-Ray.
2. If the user wants to compare the two, they wouldn't want that config set globally, and might find that confusing.

I personally think that making the user explicitly set a config to enable Dask-on-Ray is the better way to go. If you think that this configuration is getting ugly (and could get uglier if we add more Ray-specific optimization passes), we could provide a helper like `enable_dask_on_ray()` that hides the complexity of manually setting these Dask configs.

Comment (ericl) [reply doc/source/dask-on-ray.rst:91]: Makes sense. I think adding something like `ray.util.enable_dask_integration()` would be very helpful for users.

Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:91]: Agreed.

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [inline doc/source/dask-on-ray.rst:55]: Please also mention the mature actor implementation. Also, let's put this as the first reason (since I believe it is the strongest benefit).

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [inline doc/source/dask-on-ray.rst:61]: Please explain this is "currently" not supported. (We can probably starting supporting it once ray client becomes more stable)

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [reply doc/source/dask-on-ray.rst:12]: Also, it'll be great how we achieved this (mention we used Dask's plugin APIs to implement this).

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [inline doc/source/dask-on-ray.rst:48]: I think we should write more examples in a different section, but we can probably do that later.

Comment (rkooo567) [reply doc/source/dask-on-ray.rst:48]: Also, can you write a note under this section that every Dask API should be available, but if users find any issue, create an issue to our github issue page?

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [inline doc/source/dask-on-ray.rst:59]: Put this under 

```
.. note::
```
section

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [inline doc/source/dask-on-ray.rst:92]: We should mention that max_branch should be large in this case.

File: doc/source/dask-on-ray.rst
Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:73]: I got a 10x speedup on my local machine for a smaller shuffle, but we can limit it to speedups that we got on our more official benchmarks if you'd like.

File: doc/source/dask-on-ray.rst
Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:55]: That's a good motivation for using Ray in general, but Dask-on-Ray doesn't use actors, so I think that calling out such an orthogonal feature doesn't really fit here. We can point them to a more general "why use Ray" docs page that contains a more comprehensive listing of Ray's features, but I think that we should be calling out Ray features that Dask workloads will benefit from here.

File: doc/source/dask-on-ray.rst
Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:12]: I don't think that we will ever want to implement Dask Distributed's scheduler protocol and/or the client API, at least not in its current form. I think that it's more likely that we would offer our own client that offers an API that better fits the task/cluster state introspection and management capabilities and APIs that Ray exposes.

File: doc/source/dask-on-ray.rst
Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:61]: I don't think that we will ever want to implement Dask Distributed's scheduler protocol and/or the client API, at least not in its current form. I think that it's more likely that we would offer our own client that offers an API that better fits the task/cluster state introspection and management capabilities and APIs that Ray exposes.

File: doc/source/dask-on-ray.rst
Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:13]: But object spilling won't be turned on by default until the 1.3 release, right? Should we wait to call out that feature until it's on by default in the most recent release, or should we call it out and include instructions for turning it on?

Comment (rkooo567) [reply doc/source/dask-on-ray.rst:13]: I think it is fine because we already have a doc in 1.2.0 about how to turn on the object spilling. The feature is probably not as stable as 1.3.0, but we can mention the feature is under active development and report issues if they face any.

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [reply doc/source/dask-on-ray.rst:73]: Maybe 4~10X?

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [reply doc/source/dask-on-ray.rst:55]: Hmm yeah a good point about actors. But shared memory is definitely a useful feature. Should we only mention shared memory in this case?

Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:55]: Shared memory is mentioned, or are you saying that you think that shared memory should go first, before mentioning the cluster launcher?

File: doc/source/dask-on-ray.rst
Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:73]: I'll keep it to the official benchmarks and say 4x.

File: doc/source/dask-on-ray.rst
Comment (stephanie-wang) [inline doc/source/dask-on-ray.rst:73]:
```suggestion
Dask-on-Ray is an ongoing project and is not expected to achieve the same performance as using Ray directly. All `Dask abstractions <https://docs.dask.org/en/latest/user-interfaces.html>`__ should run seamlessly on top of Ray using this scheduler, so if you find that one of these abstractions doesn't run on Ray, please `open an issue <https://github.com/ray-project/ray/issues/new/choose>`__.
```

File: doc/source/dask-on-ray.rst
Comment (stephanie-wang) [reply doc/source/dask-on-ray.rst:109]: This should be pretty quick, right? @clarkzinzow, do you want to do that in this PR?

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [reply doc/source/dask-on-ray.rst:69]: Yeah I think the first reason is the least compelling reason compared to other 2. Maybe we can change the order? 2 -> 3 -> 1 or 3 -> 2 -> 1

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [inline doc/source/dask-on-ray.rst:77]: Can you align the match the length of the string and `=`? 

```
=====================
Out-of-Core Data Processing

```

File: doc/source/dask-on-ray.rst
Comment (rkooo567) [inline doc/source/dask-on-ray.rst:80]: We shouldn't tag master here directly. There's a way to make a link natively. (Use this as an example; https://github.com/ray-project/ray/blob/master/doc/examples/placement-group.rst)

**Review by rkooo567 — APPROVED** [review]
This is awesome!! A couple of last comments are left, @ericl I will merge this PR after he addressed them.

**Comment by rkooo567** [comment]
Oh, and can you also move spark on ray under the data processing section? https://github.com/ray-project/ray/pull/14018

**Comment by clarkzinzow** [comment]
@rkooo567 Ah cool I didn't know that the Spark-on-Ray docs had been merged, I'll add that!

**Comment by wuisawesome** [comment]
@clarkzinzow can you ping me when this is ready to be picked into the 1.2.0 docs?

File: doc/source/dask-on-ray.rst
Comment (clarkzinzow) [reply doc/source/dask-on-ray.rst:104]: That should be very quick, yes, but I'll keep this PR focused on docs and open another PR that adds that helper.

**Comment by rkooo567** [comment]
@wuisawesome I will merge it once the doc build passes and lyk

**Comment by richardliaw** [comment]
doc build passed

**Comment by rkooo567** [comment]
@richardliaw Would you like to have a quick review before we merge it?

File: doc/source/dask-on-ray.rst
Comment (richardliaw) [inline doc/source/dask-on-ray.rst:55]: this is a bit hard to read on the docs; try 80 column width?

File: doc/source/dask-on-ray.rst
Comment (richardliaw) [inline doc/source/dask-on-ray.rst:73]: the spacing turns out weird

```suggestion
2. If you'd like to use Dask and Ray libraries in the same application without having two different clusters.
3. If you'd like to create data analyses using the familiar NumPy and Pandas APIs provided by Dask and execute them on a fast, fault-tolerant distributed task execution system geared towards production, like Ray.
```

File: doc/source/dask-on-ray.rst
Comment (richardliaw) [inline doc/source/dask-on-ray.rst:69]: spacing turns out weird
```suggestion
1. To take advantage of Ray-specific features such as the
```

File: doc/source/dask-on-ray.rst
Comment (richardliaw) [inline doc/source/dask-on-ray.rst:115]: hard to read -- can we do 80 col width?

File: doc/source/dask-on-ray.rst
Comment (richardliaw) [inline doc/source/dask-on-ray.rst:93]:
```suggestion
Dask on Ray provides a Dask DataFrame optimizer that leverages Ray's ability to
```

File: doc/source/dask-on-ray.rst
Comment (richardliaw) [inline doc/source/dask-on-ray.rst:19]:
```suggestion
Ray provides a scheduler for Dask (`dask_on_ray`) which allows you to build data
analyses using Dask's collections and execute
the underlying tasks on a Ray cluster. 

`dask_on_ray` uses Dask's scheduler API, which allows you to
specify any callable as the scheduler that you would like Dask to use to execute your
```

Comment (richardliaw) [reply doc/source/dask-on-ray.rst:19]: (nit: maintain 3rd or 2nd person)

**Comment by richardliaw** [comment]
done @rkooo567 ! @clarkzinzow take a look?

**Comment by rkooo567** [comment]
@wuisawesome

**Pull request merged**
